core/src/test/java/org/infinispan/eviction/impl/EvictionWithPassivationTest.java

```diff
@@ -232,59 +236,88 @@ public void testActivationPutAllInBatchRolledBack() throws Exception {
          testCache.put("key" + i, "value" + i);
       }
 
+      // Eviction notifiation can be non blocking async in certain configs - so wait for notification to complete
+      evictionListener.phaser.awaitAdvanceInterruptibly(phase, 10, TimeUnit.SECONDS);
       String evictedKey = evictionListener.getEvictedKey();
-      assertTrue(isEntryInStore(evictedKey));
+      isEntryInStore(evictedKey, true);
       testCache.compute(evictedKey, (k ,v) -> v + "-modfied");
-      assertFalse(isEntryInStore(evictedKey));
+      isEntryInStore(evictedKey, false);
    }
 
    public void testRemoveViaComputeOnEvictedEntry() throws Exception {
       Cache<String, String> testCache = cacheManager.getCache(CACHE_NAME);
+      int phase = evictionListener.phaser.getPhase();
       for (int i = 0; i < EVICTION_MAX_ENTRIES + 1; i++) {
          testCache.put("key" + i, "value" + i);
       }
 
+      // Eviction notifiation can be non blocking async in certain configs - so wait for notification to complete
+      evictionListener.phaser.awaitAdvanceInterruptibly(phase, 10, TimeUnit.SECONDS);
       String evictedKey = evictionListener.getEvictedKey();
-      assertTrue(isEntryInStore(evictedKey));
+      if (evictedKey == null) {
+         System.currentTimeMillis();
+      }
+      isEntryInStore(evictedKey, true);
       testCache.compute(evictedKey, (k ,v) -> null);
       assertFalse(testCache.containsKey(evictedKey));
-      assertFalse(isEntryInStore(evictedKey));
+      isEntryInStore(evictedKey, false);
    }
 
    public void testCleanStoreOnPut() throws Exception {
       Cache<String, String> testCache = cacheManager.getCache(CACHE_NAME);
       testCache.clear();
       putIntoStore("key", "oldValue");
       testCache.put("key", "value");
-      assertFalse(isEntryInStore("key"));
+      isEntryInStore("key", false);
    }
 
-   private boolean isEntryInStore(String key) throws Exception {
+   private void isEntryInStore(String key, boolean expectPresent) throws Exception {
+      assertNotNull(key);
       Cache<String, String> testCache = cacheManager.getCache(CACHE_NAME);
-      CacheLoader<String, String> loader = TestingUtil.getFirstLoader(testCache);
-      Object loaderKey = key;
+
+      Object loaderKey;
       if (storage == StorageType.OFF_HEAP) {
          GlobalMarshaller gm = TestingUtil.extractGlobalMarshaller(testCache.getCacheManager());
          loaderKey = new WrappedByteArray(gm.objectToByteBuffer(key));
+      } else {
+         loaderKey = key;
+      }
+
+      CompletionStage<MarshallableEntry<String, String>> stage = TestingUtil.extractComponent(testCache, PersistenceManager.class)
+            .loadFromAllStores(loaderKey, true, true);
+      MarshallableEntry<String, String> entry = CompletionStages.join(stage);
+      if (expectPresent) {
+         assertNotNull(entry);
+      } else {
+         assertNull(entry);
+      }
+      CacheLoader<String, String> loader = TestingUtil.getFirstLoader(testCache);
```

Comment:
Doesn't this duplicate the `loadFromAllStores` check?

Checking that cache loaders properly implement the SPI should be done in the store tests.

Reply:
Not quite, the `loadFromAllStores` can be grabbing from the internal `PersistenceManager` map. The latter part just ensure the entry is eventually written to the store.

Reply:
Could you force the test to hit or not hit the map when it calls `loadFromAllStores()` then? Otherwise the test could always go on one branch and the other branch would remain untested.